File: src/praisonai-agents/praisonaiagents/runtime/protocols.py
"""
Runtime Protocols for PraisonAI Agents.

This module defines the protocol interfaces that runtime implementations
must implement. These protocols standardize the execution contract
across different runtime types (native, plugin, managed, etc.).

Includes both turn-context protocols and agent runtime protocols to support
different execution patterns.

Protocol-driven design following AGENTS.md:
- Lightweight protocols only (no heavy implementations)
- Dataclasses for configuration
- Async-first with proper typing
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, AsyncIterator, Dict, List, Optional, Protocol, runtime_checkable, TYPE_CHECKING

if TYPE_CHECKING:
    from .turn_context import PreparedTurnContext, RuntimeMode
    from ..agent.protocols import AgentProtocol


# Turn-context protocols from main

@runtime_checkable
class TurnRuntimeProtocol(Protocol):
    """
    Protocol for turn-based runtime execution.
    
    This protocol defines the interface that runtimes must implement to
    execute prepared turn contexts. It standardizes the execution contract
    across different runtime types (native, plugin, managed, etc.).
    
    Example:
        ```python
        class MyRuntime:
            async def run_turn(self, context: PreparedTurnContext) -> str:
                # Execute the turn using the prepared context
                return "response"
        
        # Use with prepared context
        context = PreparedTurnContext(...)
        runtime = MyRuntime()
        result = await runtime.run_turn(context)
        ```
    """
    
    async def run_turn(self, context: PreparedTurnContext) -> str:
        """
        Execute a single turn using the prepared context.
        
        This is the main execution entry point for all runtime types.
        The context is immutable and contains all necessary configuration
        for the turn execution.
        
        Args:
            context: The prepared turn context containing model, tools,
                    transcript, delivery channels, and correlation IDs
                    
        Returns:
            The agent's response as a string
            
        Raises:
            RuntimeError: If execution fails
            ValueError: If context is invalid or incompatible
        """
        ...
    
    def supports_runtime_mode(self, mode: RuntimeMode) -> bool:
        """
        Check if this runtime supports the given execution mode.
        
        Args:
            mode: The runtime mode to check
            
        Returns:
            True if the mode is supported, False otherwise
        """
        ...
    
    def get_supported_modes(self) -> List[RuntimeMode]:
        """
        Get all runtime modes supported by this implementation.
        
        Returns:
            List of supported RuntimeMode values
        """
        ...


@runtime_checkable  
class TurnContextBuilderProtocol(Protocol):
    """
    Protocol for building PreparedTurnContext instances.
    
    This protocol defines the interface for context builders that prepare
    turn contexts from agent configuration and request parameters.
    """
    
    def build_context(
        self,
        agent: AgentProtocol,
        prompt: str,
        **kwargs: Any
    ) -> PreparedTurnContext:
        """
        Build a prepared turn context from agent and request.
        
        Args:
            agent: The agent instance
            prompt: The user prompt for this turn
            **kwargs: Additional request parameters
            
        Returns:
            A prepared turn context ready for execution
        """
        ...


# Agent runtime protocols and dataclasses from this branch

@dataclass
class RuntimeConfig:
    """Declarative runtime configuration.
    
    Base configuration class for runtime implementations.
    Specific runtimes can extend this for their own config needs.
    """
    # Core runtime identification
    runtime_id: str
    
    # Optional metadata
    metadata: Dict[str, Any] = field(default_factory=dict)


@dataclass
class RuntimeResult:
    """Result from runtime execution."""
    content: str
    metadata: Dict[str, Any] = field(default_factory=dict)
    error: Optional[str] = None


@dataclass
class RuntimeDelta:
    """Streaming delta from runtime execution."""
    type: str  # "text" | "tool_call" | "thinking" | "error"
    content: str = ""
    metadata: Dict[str, Any] = field(default_factory=dict)


@runtime_checkable
class AgentRuntimeProtocol(Protocol):
    """Protocol for agent runtime implementations.
    
    Any object implementing these methods can serve as an Agent runtime.
    Follows typing.Protocol pattern from AGENTS.md.
    
    This is a simpler contract than CliBackendProtocol, focused on 
    the minimal interface needed for runtime execution.
    """
    
    def supports(self, model_ref: Optional[str] = None) -> bool:
        """Check if this runtime supports the given model reference.
        
        Args:
            model_ref: Optional model reference (e.g., "gpt-4", "claude-3")
            
        Returns:
            True if this runtime can handle the model, False otherwise
        """
        ...
    
    async def run_turn(
        self, 
        prompt: str, 
        *,
        system_prompt: Optional[str] = None,
        model_ref: Optional[str] = None,
        **kwargs
    ) -> RuntimeResult:
        """Execute a single turn and return result.
        
        Args:
            prompt: User prompt/query
            system_prompt: Optional system prompt
            model_ref: Optional model reference
            **kwargs: Additional runtime-specific options
            
        Returns:
            RuntimeResult with response content and metadata
        """
        ...
    
    async def stream_turn(
        self, 
        prompt: str, 
        **kwargs
    ) -> AsyncIterator[RuntimeDelta]:
        """Stream response deltas from runtime execution.
        
        Args:
            prompt: User prompt/query
            **kwargs: Additional options (system_prompt, model_ref, etc.)
            
        Yields:
            RuntimeDelta objects with incremental response content
        """
        ...


@runtime_checkable
class StreamingRuntimeProtocol(Protocol):
    """Extended protocol for runtimes with enhanced streaming capabilities.
    
    Optional protocol that runtimes can implement to indicate support for
    true incremental streaming. This will be the migration path when the
    underlying Agent implementation adds AsyncIterator support.
    
    Example:
        ```python
        # Check if runtime supports enhanced streaming
        if isinstance(runtime, StreamingRuntimeProtocol):
            if runtime.supports_incremental_streaming():
                # Use enhanced streaming
                async for delta in runtime.stream_turn_incremental(prompt):
                    process_delta(delta)
        ```
    """
    
    def supports_incremental_streaming(self) -> bool:
        """Check if this runtime supports true incremental streaming.
        
        Returns:
            True if the runtime can stream incremental deltas,
            False if it only supports single-delta responses
        """
        ...
    
    async def stream_turn_incremental(
        self,
        prompt: str,
        **kwargs
    ) -> AsyncIterator[RuntimeDelta]:
        """Stream true incremental response deltas.
        
        This method will be called when true streaming is available.
        Unlike the base stream_turn which may return a single delta,
        this guarantees incremental streaming of response chunks.
        
        Args:
            prompt: User prompt/query
            **kwargs: Additional options (system_prompt, model_ref, etc.)
            
        Yields:
            RuntimeDelta objects with incremental response chunks
        """
        ...

File: src/praisonai-ts/src/runtime/protocols.ts
/**
 * Runtime protocols for PraisonAI agents.
 *
 * Port of `praisonaiagents/runtime/protocols.py`. Defines the interfaces a
 * runtime implementation must satisfy so execution is standardised across
 * runtime types (native, plugin, managed, ...).
 *
 * Not ported here (documented gaps):
 * - `runtime/turn_context.py` (`PreparedTurnContext`) — the turn-context
 *   protocols below are generic over the context type instead.
 * - `runtime/capabilities.py` (`RuntimeCapabilityMatrix`) — represented as an
 *   open record ({@link RuntimeCapabilityMatrix}) until that module is ported.
 *
 * Python `@runtime_checkable` `isinstance` checks map to the structural
 * guard {@link isAgentRuntime}.
 */

/** Execution mode for the runtime (Python `turn_context.py::RuntimeMode`). */
export type RuntimeMode = 'sync' | 'async' | 'stream' | 'async_stream';

/**
 * Protocol for turn-based runtime execution
 * (Python `protocols.py::TurnRuntimeProtocol`).
 *
 * @typeParam TContext The prepared turn context type (Python `PreparedTurnContext`).
 */
export interface TurnRuntimeProtocol<TContext = unknown> {
  /** Execute a single turn using the prepared context; resolves to the agent's response. */
  runTurn(context: TContext): Promise<string>;
  /** Check if this runtime supports the given execution mode. */
  supportsRuntimeMode(mode: RuntimeMode): boolean;
  /** All runtime modes supported by this implementation. */
  getSupportedModes(): RuntimeMode[];
}

/**
 * Protocol for building prepared turn contexts
 * (Python `protocols.py::TurnContextBuilderProtocol`).
 */
export interface TurnContextBuilderProtocol<TAgent = unknown, TContext = unknown> {
  buildContext(agent: TAgent, prompt: string, kwargs?: Record<string, unknown>): TContext;
}

/** Constructor options for {@link RuntimeConfig}. */
export interface RuntimeConfigInit {
  /** Core runtime identification. */
  runtimeId: string;
  /** Optional metadata. Default `{}`. */
  metadata?: Record<string, unknown>;
}

/**
 * Declarative runtime configuration (Python `protocols.py::RuntimeConfig`).
 */
export class RuntimeConfig {
  runtimeId: string;
  metadata: Record<string, unknown>;

  constructor(config: RuntimeConfigInit) {
    this.runtimeId = config.runtimeId;
    this.metadata = config.metadata ?? {};
  }
}

/** Constructor options for {@link RuntimeResult}. */
export interface RuntimeResultInit {
  content: string;
  /** Default `{}`. */
  metadata?: Record<string, unknown>;
  /** Default `null`. */
  error?: string | null;
}

/** Result from runtime execution (Python `protocols.py::RuntimeResult`). */
export class RuntimeResult {
  content: string;
  metadata: Record<string, unknown>;
  error: string | null;

  constructor(config: RuntimeResultInit) {
    this.content = config.content;
    this.metadata = config.metadata ?? {};
    this.error = config.error ?? null;
  }
}

/** Streaming delta kinds (Python comment on `RuntimeDelta.type`). */
export type RuntimeDeltaType = 'text' | 'tool_call' | 'thinking' | 'error';

/** Constructor options for {@link RuntimeDelta}. */
export interface RuntimeDeltaInit {
  type: RuntimeDeltaType | string;
  /** Default `""`. */
  content?: string;
  /** Default `{}`. */
  metadata?: Record<string, unknown>;
}

/** Streaming delta from runtime execution (Python `protocols.py::RuntimeDelta`). */
export class RuntimeDelta {
  type: RuntimeDeltaType | string;
  content: string;
  metadata: Record<string, unknown>;

  constructor(config: RuntimeDeltaInit) {
    this.type = config.type;
    this.content = config.content ?? '';
    this.metadata = config.metadata ?? {};
  }
}

/**
 * Capability matrix reported by a runtime. Placeholder for
 * `runtime/capabilities.py::RuntimeCapabilityMatrix`, which is not yet
 * ported; an open record keeps the protocol shape stable until it is.
 */
export type RuntimeCapabilityMatrix = Record<string, unknown>;

/**
 * Options for {@link AgentRuntimeProtocol.runTurn}: Python's keyword-only
 * `system_prompt`, `model_ref` plus `**kwargs`.
 */
export interface RunTurnOptions {
  /** Optional system prompt. Default `null`. */
  systemPrompt?: string | null;
  /** Optional model reference. Default `null`. */
  modelRef?: string | null;
  /** Additional runtime-specific options (Python `**kwargs`). */
  [key: string]: unknown;
}

/**
 * Protocol that all agent runtime implementations must follow
 * (Python `protocols.py::AgentRuntimeProtocol`).
 *
 * Combines turn-based execution with capability reporting so a runtime can
 * be both executed and validated for compatibility.
 */
export interface AgentRuntimeProtocol {
  /** Human-readable name, e.g. `"native"`, `"claude-code"` (Python property `runtime_name`). */
  readonly runtimeName: string;
  /** Version string used for compatibility tracking (Python property `runtime_version`). */
  readonly runtimeVersion: string;
  /** Report the capability matrix for this runtime. */
  capabilities(): RuntimeCapabilityMatrix;
  /** Check if this runtime supports the given model reference. */
  supports(modelRef?: string | null): boolean;
  /** Execute a single turn and return the result. */
  runTurn(prompt: string, options?: RunTurnOptions): Promise<RuntimeResult>;
  /** Stream response deltas from runtime execution. */
  streamTurn(prompt: string, kwargs?: Record<string, unknown>): AsyncIterable<RuntimeDelta>;
  /** Execute an agent with the given configuration and prompt. */
  executeAgent(
    agentConfig: Record<string, unknown>,
    prompt: string,
    kwargs?: Record<string, unknown>,
  ): Promise<Record<string, unknown>>;
  /** Execute an agent with streaming responses (only required with the streaming capability). */
  streamAgent(
    agentConfig: Record<string, unknown>,
    prompt: string,
    kwargs?: Record<string, unknown>,
  ): AsyncIterable<unknown>;
  /** Validate agent configuration; resolves to error messages (empty if valid). */
  validateConfig(agentConfig: Record<string, unknown>): Promise<string[]>;
  /** Perform a runtime health check. */
  healthCheck(): Promise<Record<string, unknown>>;
}

const AGENT_RUNTIME_METHODS: ReadonlyArray<keyof AgentRuntimeProtocol> = [
  'capabilities',
  'supports',
  'runTurn',
  'streamTurn',
  'executeAgent',
  'streamAgent',
  'validateConfig',
  'healthCheck',
];

/**
 * Structural check equivalent to Python's
 * `isinstance(value, AgentRuntimeProtocol)` on the `@runtime_checkable`
 * protocol: every method must be present and `runtimeName` /
 * `runtimeVersion` must be strings.
 */
export function isAgentRuntime(value: unknown): value is AgentRuntimeProtocol {
  if (value === null || (typeof value !== 'object' && typeof value !== 'function')) return false;
  const candidate = value as Record<string, unknown>;
  if (typeof candidate.runtimeName !== 'string' || typeof candidate.runtimeVersion !== 'string') {
    return false;
  }
  return AGENT_RUNTIME_METHODS.every((method) => typeof candidate[method] === 'function');
}
